### DISCUSSION TITLE
Implementation of sift algorithm

### DISCUSSION THREAD

**Discussion by wujilingfeng** [AccelerateHS/accelerate#527] [Show and tell]
Recently, I implemented the sift algorithm (all its core algorithms have been implemented, just short of some basic assembly), and opened it up [haskell-sift](https://gitee.com/wujilingfeng/haskell-sift).

It still needs more professional optimization, because it is not as fast as opencv under the same CPU acceleration.
It would be better if awhile function could be optimized. 

Due to the limited personal level, if anyone can provide professional guidance, this project will be beneficial to everyone.

**Comment by tmcdonell** [comment]
oh nice! I haven't looked in detail yet, but am curious to try it out.
What do you mean by optimising the `awhile` function?

**Reply by wujilingfeng** [reply]
```

compute_key_point_directions:: Acc (Array DIM3 Float)->Acc( Vector (Float,Float,Float))->Acc (Vector (Float,Float,Float,Float,Float))
compute_key_point_directions gauss_octave feas = A.afst loop  
  where
    A.I1 n=A.shape feas
    initial_ =(use $ A.fromList (Z:.0) [] )::Acc (Vector (Float,Float,Float,Float,Float))
    loop= A.awhile 
      (\(A.T2 _ s)-> A.map (A.< n) s ) 
      (\(A.T2 f s )->
        let 
          ss=the s
          new= compute_key_point_direction gauss_octave (feas A.! (A.I1 ss))
        in 
          A.T2  (f A.++ new )  (unit (ss+1))  ) 
      (A.T2  initial_ (A.unit 0))
```
the code is like above. compute_key_point_direction‘s function signature is
```
compute_key_point_direction::Acc (Array DIM3 Float)->Exp (Float,Float,Float)-> Acc (Vector (Float,Float,Float,Float,Float))

```
Awhile function is mainly used to splice arrays, and it consume a lot of time

**Reply by tmcdonell** [reply]
Interesting. I think I have seen this pattern before so have been thinking about it a bit. It would be good for us to provide some way to support this better (without copying).

**Reply by tmcdonell** [reply]
The example I had previously worked a bit differently, adding one row to a matrix at a time rather than appending everything to a single vector. Good to have something different to think about...

**Reply by tmcdonell** [reply]
I wonder if the [`expand`](https://github.com/AccelerateHS/accelerate/blob/master/src/Data/Array/Accelerate/Prelude.hs#L2516) operation could be used here?

**Reply by wujilingfeng** [reply]
The expand's signature is `(Exp a->Exp Int)->(Exp a->Exp Int->Exp b)->Acc (Vector a)->Acc (Vector b)`,
But `(Exp a->Exp Int->Exp b)`  is difficult to write.

**Reply by wujilingfeng** [reply]
I think awhileis is serial, resulting in slow speed.
If only there were such an operator like ` (Exp a->Acc (Vector a))->Acc (Vector a)->Acc (Vector a )`, And ensure that it is parallel

**Reply by wujilingfeng** [reply]
Is the concurrency strategy feasible? like `rpar`? 
-------
Not work.

**Reply by tmcdonell** [reply]
The `awhile` is sequential yes. And each iteration you use `(++)` to append the new result to the accumulator from the previous stop, which is going to copy the data, which is slow. But each loop iteration is independent, so you *should* (at least in principle) be able to do all of those iterations in parallel at once.

So a first question is, does `compute_key_point_direction` always return an array of the same length? A quick look over the code seems to indicate that this is probably not true, since I see a filter in there. It might still be possible however since `filter` can take a 2D array and will return a vector. But there is a lot going on in that function so I'm not sure...

Another option would be, that this would also be much faster if `permute` (more reliably) did in-place updates (it does sometimes, but I don't think it will in your case). Then you could allocate a big vector at front and each step of the loop just directly fills in the next `n` elements. We don't have something like that at the moment, but @ivogabe has been thinking about this.

Separately, @NOerlemans and I have been discussing a different kind of abstraction that might help you (sort of closer to the `rpar` you wish for) but that is a long way off...

**Reply by wujilingfeng** [reply]
> ‘’‘Then you could allocate a big vector at front and each step of the loop just directly fills in the next n elements. We don't have something like that at the moment, but @ivogabe has been thinking about this’‘’

 This idea is very useful.

**Reply by wujilingfeng** [reply]
I optimized the algorithm and replaced `awhile` function with `expand`. There are two places that need to be replaced。
This work requires algorithm innovation and is not easy.Later, I will update the code warehouse. But the memory will be exhausted, and the program will exit, unable to complete the complete calculation.
I think it is more important to provide better operators than to think about creative algorithms.

**Reply by wujilingfeng** [reply]
By the way, opencv is optimized very well. 
Can the code written using the accelerate library be faster on GPU than opencv?I look forward to trying in the future.

It seems that opencv uses instruction set optimization，
It seems that the speed of instruction set optimization is extremely fast！

**Reply by tmcdonell** [reply]
Being able to express nested parallelism would also make this much simpler, then you could just sort of `concatMap compute_key_point_direction`. We had an old sequences branch that could also do that but it's quite bit-rotted at this point.

What kinds of optimisations does OpenCV do?

**Reply by wujilingfeng** [reply]
According to the information I collected, opencv may use instruction set optimization.

By the way，
 Why is  `permute` defined like this `(Exp a ->Exp a->Exp a)->Acc (Array sh' a)->(Exp sh->Exp (Maybe sh'))->Acc (Array sh a)->Acc (Array sh' a)` rather than `(Exp b ->Exp a->Exp a)->Acc (Array sh' a)->(Exp sh->Exp (Maybe sh'))->Acc (Array sh b)->Acc (Array sh' a)`?

**Reply by wujilingfeng** [reply]
I have updated the code, now it's very fast. The memory usage is still too high

**Reply by tomsmeding** [reply]
> By the way， Why is `permute` defined like this `(Exp a ->Exp a->Exp a)->Acc (Array sh' a)->(Exp sh->Exp (Maybe sh'))->Acc (Array sh a)->Acc (Array sh' a)` rather than `(Exp b ->Exp a->Exp a)->Acc (Array sh' a)->(Exp sh->Exp (Maybe sh'))->Acc (Array sh b)->Acc (Array sh' a)`?

Because of the way `permute` is implemented. It shoots off an atomic update for every cell in the source array, and there are no guarantees whatsoever in which order those will be executed. That is also why the combination function must usually be commutative and associative for `permute` to give a well-defined, consistent answer. (If all updates are at distinct indices then these properties are not required, because then there is nothing to reorder.)

**Reply by wujilingfeng** [reply]
if `permute`  is defined `(Exp b ->Exp a->Exp a)->Acc (Array sh' a)->(Exp sh->Exp (Maybe sh'))->Acc (Array sh b)->Acc (Array sh' a)`, I can solve the memory problem.

**Reply by wujilingfeng** [reply]
If I want to implement histogram, what other good ways are there?

**Reply by wujilingfeng** [reply]
I've solved the memory problem and now it's very fast. Thank you for your guidance and suggestions. I will update the code later. I am also trying to install the accelerate library in Windows.
After that, I will also add visual programs, hoping that this project can bring benefits to everyone.
Thank you！
